### PR TITLE
Add pathlib support

### LINKED ISSUE #6: Add support for pathlib Paths
**Is your feature request related to a problem? Please describe.**
A description of what the problem is. Ex. What are you trying to solve?

After dropping support for Python 3.2-3.5 we can now take advantage of `pathlib` in all its glory.

**Describe the solution you'd like**
A detailed description of an ideal implementation.

Current path arguments that expect a `str` should be modified to accept `str` or `Path`, and os.path calls replaced.

### PR DESCRIPTION
**Pull request by parafoxia**
### Summary
This PR modifies the API to utilise the power of `pathlib`. It also fixes some typehinting errors Pyright didn't pick up. There's a new Mypy test within the noxfile, though the Pyright one has been retained.

### Tasks
- [x] I have searched for similar/duplicate PR's, and there are none.
- [x] I have made tests to validate any changes to the code base.
- [x] I have run `nox` and all sessions pass.

### Related issues
Closes #6.

### CODE REVIEW

**Comment by parafoxia** [comment]
Do we think this should be released as v0.3.2 or v0.4.0? Part of me is leaning to the latter cos there's a lot of changes, but another part toward the former because all the changes are in the backend, and there aren't really any "new" features.

**Comment by Jonxslays** [comment]
> Do we think this should be released as v0.3.2 or v0.4.0? Part of me is leaning to the latter cos there's a lot of changes, but another part toward the former because all the changes are in the backend, and there aren't really any "new" features.

According to [SemVer](https://semver.org/) we would be ok with either option, if you feel 0.4 is warranted lets go with that.

> Minor version Y (x.Y.z | x > 0) MAY be incremented if substantial new functionality or improvements are introduced within the private code. It MAY include patch level changes.

**Comment by Jonxslays** [comment]
But to be fair, we still haven't released 1.0 so I don't think its too relevant right now.

File: len8/models/checker.py [hunk 1]
```diff
@@ -93,114 +110,102 @@ def extend(self, extend: bool) -> None:
...
-                raise errors.InvalidPath(p)
+        for e in self.exclude:
+            if path.is_relative_to(e):
+                return False
 
-            if os.path.isfile(p):
-                self._check_file(p)
-            else:
-                self._check_dir(p)
+        return True
```
Comment (Jonxslays) [inline len8/models/checker.py:133]: This change does not take into account someone passing an absolute path to the file they want to exclude. 

Assume every file has lines that are too long for minimal reproduction:

Directory structure
```bash
.
├── ignore_me
│   ├── test2.py
│   └── test.py
└── test.py
```

When using a relative path, `./test.py` is ignored as expected.
```bash
$ len8 . -x test.py
2 line(s) are too long:
- /home/jon/repro/ignore_me/test.py, line 1 (119/79)
- /home/jon/repro/ignore_me/test2.py, line 1 (119/79)
```

When using an absolute path for the exclude, `test.py` is not ignored.
```bash
$ len8 . -x $(pwd)/test.py
3 line(s) are too long:
- /home/jon/repro/test.py, line 1 (119/79)
- /home/jon/repro/ignore_me/test.py, line 1 (119/79)
- /home/jon/repro/ignore_me/test2.py, line 1 (119/79)

$ echo $(pwd)/test.py
/home/jon/repro/test.py
```

**Review by Jonxslays — CHANGES_REQUESTED** [review]
Just a couple of things here.

**Review by Jonxslays — APPROVED** [review]
amazing, you cleaned it up a lot.

**Pull request merged**